### PR TITLE
Feature/fingerbank nba

### PR DESCRIPTION
**Pull request by julsemaan**
# Missing
* Documentation

# Description
Adds support for network behavior analysis (anomaly detection)

# Impacts
Fingerbank integration

# Code / PR Dependencies
https://git.inverse.ca/inverse/fingerbank-collector/pulls/29
https://github.com/fingerbank/perl-client/pull/47

# NEW Package(s) required
fingerbank-4.2.0
fingerbank-collector-1.3.0

# Delete branch after merge
YES

# Checklist
(REQUIRED) - [yes, no or n/a]
- [ ] Document the feature
- [ ] Add unit tests
- [ ] Add acceptance tests (TestLink)

# NEWS file entries
(REQUIRED, but may be optional...)
## New Features
* Added support for anomaly detection through Fingerbank

### CODE REVIEW

File: conf/network_behavior_policies.conf.example [hunk 1]
```diff
@@ -0,0 +1,23 @@
+[android_blacklisted]
+blacklisted_ip_hosts_threshold=1
+device_attributes_diff_score=0
+whitelisted_ips=1.1.1.1/32,8.8.8.8/32
```
Comment (fdurand) [inline conf/network_behavior_policies.conf.example]: 8.8.4.4/32

File: html/pfappserver/root/static.alt/src/views/Configuration/_config/networkBehaviorPolicy.js [hunk 2]
```diff
@@ -0,0 +1,380 @@
...
+              attrs: {
+                values: { checked: 'enabled', unchecked: 'disabled' }
+              },
+              validators: validatorsFromMeta(meta, 'watch_blacklisted_ips', 'Watch blacklisted IPs')
+            }
+          ]
+        },
+        {
+          label: i18n.t('Whitelisted IPs'),
+          text: i18n.t('Which IPs (can be CIDR) to ignore when checking against the blacklisted IPs list'),
```
Comment (fdurand) [inline html/pfappserver/root/static.alt/src/views/Configuration/_config/networkBehaviorPolicy.js]: comma delimited

File: lib/pf/ConfigStore/NetworkBehaviorPolicy.pm [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+
+=head2 cleanupBeforeCommit
+
+Clean data before update or creating
+
+=cut
+
+sub cleanupBeforeCommit {
+    my ($self, $id, $data) = @_;
+    use Data::Dumper;print Dumper($data);
```
Comment (fdurand) [inline lib/pf/ConfigStore/NetworkBehaviorPolicy.pm]: Remove

File: html/pfappserver/root/static.alt/src/views/Configuration/_components/NetworkBehaviorPoliciesList.vue [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+  <b-card no-body>
+    <pf-config-list
+      ref="pfConfigList"
+      :config="config"
+    >
+      <template v-slot:pageHeader>
+        <b-card-header>
+          <h4 class="mb-0">
+            {{ $t('Network Behavior Policy') }}
+            <pf-button-help class="ml-1" url="PacketFence_Installation_Guide.html#_devices_registration" />
```
Comment (cgx) [inline html/pfappserver/root/static.alt/src/views/Configuration/_components/NetworkBehaviorPoliciesList.vue]: Probably not pointing to the proper section of the documentation.

File: html/pfappserver/root/static.alt/src/views/Configuration/_config/networkBehaviorPolicy.js [hunk 5]
```diff
@@ -0,0 +1,380 @@
...
+import {
+  and,
+  not,
+  conditional,
+  hasNetworkBehaviorPolicies,
+  networkBehaviorPolicyExists
+} from '@/globals/pfValidators'
+import {
+  required
+} from 'vuelidate/lib/validators'
```
Comment (cgx) [inline html/pfappserver/root/static.alt/src/views/Configuration/_config/networkBehaviorPolicy.js]: There's currently no validation so these imports are useless.

**Comment by julsemaan** [comment]
fixed and rebased, ready for another review

File: html/pfappserver/root/static.alt/src/views/Configuration/_config/networkBehaviorPolicy.js [hunk 6]
```diff
@@ -0,0 +1,372 @@
...
+        },
+        {
+          label: 'Devices Included',
+          text: i18n.t('The list of Fingerbank devices that will be impacted by this Network Behavior Policy. Devices of this list implicitely includes all the childs of the selected devices. Leaving this empty will have all devices impacted by this policy.'),
+          cols: [
+            {
+              namespace: 'devices_included',
+              component: pfFormChosen,
+              attrs: attributesFromMeta(meta, 'devices_included'),
+              validators: validatorsFromMeta(meta, 'devices_included', 'Device')
```
Comment (cgx) [inline html/pfappserver/root/static.alt/src/views/Configuration/_config/networkBehaviorPolicy.js]: Move all the validators to the currently empty `validators` function.

**Comment by julsemaan** [comment]
Adjusted based on @cgx's comments. Will just need to do a couple of things based on previous feedback submitted by @extrafu.

Will notify here when its ready for the final review + merge

**Comment by julsemaan** [comment]
The code has been adjusted based off the comments.

We're only missing the doc but that will need to be done with the help+feedback of @extrafu 

Shall we merge this and do the doc afterwards?

File: html/pfappserver/lib/pfappserver/Form/Config/NetworkBehaviorPolicy.pm [hunk 7]
```diff
@@ -0,0 +1,180 @@
...
+        "upnp_server_strings" => "UPnP Server Strings",
+        "upnp_user_agents" => "UPnP User-Agents",
+        "user_agents" => "HTTP User-Agents",
+    );
+}
+
+
+=head1 COPYRIGHT
+
+Copyright (C) 2005-2019 Inverse inc.
```
Comment (fdurand) [inline html/pfappserver/lib/pfappserver/Form/Config/NetworkBehaviorPolicy.pm]: Copyright (C) 2005-2020 Inverse inc.

File: html/pfappserver/lib/pfappserver/Form/Field/NetworkBehaviorPolicyAttributeWeight.pm [hunk 8]
```diff
@@ -0,0 +1,101 @@
...
+=cut
+
+sub action_deflate {
+    my ($self, $value) = @_;
+    return $value->{type}.":".$value->{value};
+}
+
+=head1 COPYRIGHT
+
+Copyright (C) 2005-2019 Inverse inc.
```
Comment (fdurand) [inline html/pfappserver/lib/pfappserver/Form/Field/NetworkBehaviorPolicyAttributeWeight.pm]: Copyright (C) 2005-2020 Inverse inc.

File: lib/pf/ConfigStore/NetworkBehaviorPolicy.pm [hunk 9]
```diff
@@ -0,0 +1,87 @@
...
+    return qw(watched_device_attributes device_attributes_diff_threshold_overrides devices_included devices_excluded);
+}
+
+=head1 AUTHOR
+
+Inverse inc. <info@inverse.ca>
+
+=head1 COPYRIGHT
+
+Copyright (C) 2005-2019 Inverse inc.
```
Comment (fdurand) [inline lib/pf/ConfigStore/NetworkBehaviorPolicy.pm]: Copyright (C) 2005-2020 Inverse inc.

File: lib/pf/UnifiedApi/Controller/Config/NetworkBehaviorPolicies.pm [hunk 10]
```diff
@@ -0,0 +1,59 @@
...
+use pfappserver::Form::Config::NetworkBehaviorPolicy;
+
+ 
+=head1 AUTHOR
+
+Inverse inc. <info@inverse.ca>
+
+=head1 COPYRIGHT
+
+Copyright (C) 2005-2019 Inverse inc.
```
Comment (fdurand) [inline lib/pf/UnifiedApi/Controller/Config/NetworkBehaviorPolicies.pm]: Copyright (C) 2005-2020 Inverse inc.

**Comment by fdurand** [comment]
I am not against merging it now.

**Comment by julsemaan** [comment]
I adjusted the copyrights

**Pull request merged**